### PR TITLE
Allow stiff solver AD in adjoints

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/DiffEqSensitivity.jl/issues/198, fixes https://github.com/SciML/DiffEqFlux.jl/issues/31, fixes https://github.com/SciML/DiffEqFlux.jl/issues/271

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
We do. This just adds fallbacks where, if the caches don't match types, whatever we make a new tape and do it. I'll open an issue about adding caches for those fallback parts, but getting something working is higher priority and, given the current state of AD, probably doesn't make all that much of a difference to the final time.

**Comment by YingboMa** [comment]
Yes, for sure.

File: test/local_sensitivity/stiff_adjoints.jl [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    Hairer42(),
+    Kvaerno4(),
+    KenCarp4(),
+    Kvaerno5(),
+    KenCarp5(),
+    # Fully-Implicit Runge-Kutta Methods (FIRK)
+    RadauIIA5(),
+    # Fully-Implicit Runge-Kutta Methods (FIRK)
+    PDIRK44(),
+    # Rosenbrock Methods (all fail)
```
Comment (YingboMa) [inline test/local_sensitivity/stiff_adjoints.jl:168]: Do Rosenbrock methods still fail after the fix?

Comment (ChrisRackauckas) [reply test/local_sensitivity/stiff_adjoints.jl:168]: nope, they all work now.

Comment (ChrisRackauckas) [reply test/local_sensitivity/stiff_adjoints.jl:168]: that is just copy-pasted from an issue.

Comment (YingboMa) [inline test/local_sensitivity/stiff_adjoints.jl:168]:
```suggestion
```

File: test/local_sensitivity/stiff_adjoints.jl [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    Veldd4(),
+    Velds4(),
+    GRK4T(),
+    GRK4A(),
+    Ros4LStab(),
+    Rodas4(),
+    Rodas42(),
+    Rodas4P(),
+    Rodas5(),
+    # Rosenbrock-W Methods (all fail)
```
Comment (YingboMa) [inline test/local_sensitivity/stiff_adjoints.jl:181]:
```suggestion
    # Rosenbrock-W Methods
```

**Comment by ChrisRackauckas** [comment]
It passed but just went over time. I'll merge and stuff so that everything downstream is fixed, but add a new test group for the really long stiff solver tests that cover like every method.

**Pull request merged**